myia/analyze/value.py

```diff
@@ -0,0 +1,131 @@
+"""Value inference."""
+from typing import Dict, Callable
+from .graph import Plugin
+
+from myia import primops as P
+from myia.primops import Primitive
+from myia.utils import Named
+from myia.anf_ir import Constant, Apply, Parameter, Graph
+from myia.py_implementations import implementations as py_implementations
+
+ESTIMATORS: Dict[Primitive, Callable] = dict()
+
+
+NO_VALUE = Named('NO_VALUE')
+NOT_CONSTANT = Named('NOT_CONSTANT')
+
+
+def register_estimator(prim: Primitive):
+    """Register an estimator for a primitive.
+
+    This serves to estimate the value of a primitive in the present of
+    incomplete information (not all arguments are constants).  In some
+    cases we can know the value (like mul(0, x) -> 0).  If you
+    can't do anything unless you know all the arguments don't
+    implement this.
+    """
+    def deco(e: Callable):
+        assert prim not in ESTIMATORS
+        ESTIMATORS[prim] = e
+        return e
+    return deco
+
+
+# SCCP
+class ValuePlugin(Plugin):
+    """Implementation of SCCP (Sparse Conditional Constant Propagation).
+
+    This has some trouble with loops and such since those are
+    implemented using recursive calls and we don't handle that.
+
+    It does deal with conditionals ok.
+    """
+
+    NAME = "value"
+
+    def __init__(self, *, implementations: Dict[Primitive, Callable] = None,
+                 estimators: Dict[Primitive, Callable] = None) -> None:
```

Comment:
Couldn't we set `py_implementations` and `ESTIMATORS` as default values to the respective parameters, instead of `None`? Saves four lines of code.

Reply:
I have this ingrained aversion to mutable containers as default values.  This is because python will return the same object each time which links the calls.